### ISSUE TITLE
Recommended response behavior when identity encoding is unacceptable

### ISSUE THREAD

**Issue by gibson042** [httpwg/http-core#980]
[RFC 2616](https://www.rfc-editor.org/rfc/rfc2616.html#section-14.3) specified that identity content-coding could be made unacceptable (e.g., indicating very strong preference for response compression) by means of an Accept-Encoding request header field value like `identity;q=0` or `*;q=0` without an identity-specific preference, and then recommended that the server SHOULD respond with 406 (Not Acceptable) when it cannot send an acceptable response. But [RFC 7231](https://www.rfc-editor.org/rfc/rfc7231.html#section-5.3.4) and the [current draft](https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#rfc.section.12.5.3.p.9) update that last point to "SHOULD send a response without any content-coding", presumably primarily intended to avoid sending 406 (Not Acceptable) when the acceptability of identity content-coding is unspecified. The text still includes a description of when identity content-coding is unacceptable and still **accommodates** a "compression is mandatory" interpretation of `Accept-Encoding: gzip, identity;q=0`, but seems to strongly discourage it in the sense of [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html) ("_there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course_").

I suspect this to be unintentional, and would propose clarification in semantics:
```patch
    An Accept-Encoding header field with a field value that is empty
    implies that the user agent does not want any content coding in response.
    If an Accept-Encoding header field is present in a request and none of the
    available representations for the response have a content coding that
    is listed as acceptable, the origin server &SHOULD; send a response
-   without any content coding.
+   without any content coding unless absence of content coding is itself
+   unacceptable (in which case the origin server &SHOULD; honor that
+   preference).
```

In the absence of such clarification, the text describing when identity content-coding is unacceptable seems to be largely pointless.

**Comment by reschke** [comment]
I agree that this is not optimal, but the proposed text would change the protocol from what is has been since RFC 7231.

Given that the current draft is approved and in the publication queue, I don't think that a change is possible or even desirable.

**Comment by royfielding** [comment]
The move away from sending 406 was intentional in RFC7231. There is no implementation support for identity;q=0 overriding basic interoperability (200 response without encoding), unlike Accept for unwanted media types, because encodings are often applied late in the request handling process based on the server's analysis of whether they are efficient/possible. Hence, clients simply list the acceptable encodings and variance in the field values cause unwanted cache misses. The feature only made sense for non-compression encodings.

This does not prevent a server from responding with 406, but doing so is less interoperable (without some additional knowledge about why the client is asking for that) than sending no encoding.

**Comment by gibson042** [comment]
If it is indeed intentional for absence of content coding to always be considered acceptable, then what is the purpose of keeping the following text?
> If the representation has no content coding, then it is acceptable by default _unless specifically excluded by the Accept-Encoding header field stating either "identity;q=0" or "*;q=0" without a more specific entry for "identity"_.

**Comment by royfielding** [comment]
I meant that the move away from 406 was intentional. Suggesting that "identity;q=0" not be respected was not intentional, but is also not implemented AFAIK (It only makes sense for encryption-related encodings).

I would suggest a less normative fix, like

> An Accept-Encoding header field with a field value that is empty implies that the user agent does not want any content coding in response. If a ***non-empty*** Accept-Encoding header field is present in a request and none of the available representations for the response have a content coding that is acceptable, the origin server &SHOULD; send a response without any content coding ***unless the identity coding is indicated as unacceptable***.


and just send that now to the RFC editor as a clarification.

**Comment by gibson042** [comment]
Great, thanks!
